File: weather-bot/polymarket/src/get-balance.ts
import { createPublicClient, http, formatUnits } from "viem";
import { polygon } from "viem/chains";
import { CONFIG } from "./config";
import { logger } from "./logger";
import { errorMessage } from "./types";
import { USDC_ADDRESS, ERC20_BALANCE_ABI } from "./constants";

const SPURIOUS_ZERO_THRESHOLD_USD = 10;
const ZERO_CONFIRM_COUNT = 3; // Accept $0 as real after N consecutive zero readings
let lastKnownBalance = -1;
let consecutiveZeros = 0;
const publicClient = createPublicClient({ chain: polygon, transport: http(CONFIG.rpcUrl) });

export async function getUsdcBalance(): Promise<number> {
  try {
    const raw = await publicClient.readContract({
      address: USDC_ADDRESS,
      abi: ERC20_BALANCE_ABI,
      functionName: "balanceOf",
      args: [CONFIG.proxyWallet as `0x${string}`],
    });
    const balance = parseFloat(formatUnits(raw as bigint, 6));

    if (balance === 0 && lastKnownBalance > SPURIOUS_ZERO_THRESHOLD_USD) {
      consecutiveZeros++;
      if (consecutiveZeros >= ZERO_CONFIRM_COUNT) {
        // Confirmed real zero — accept it
        logger.warn(`Balance confirmed $0 after ${consecutiveZeros} consecutive readings`);
        lastKnownBalance = 0;
        return 0;
      }
      logger.warn(`RPC returned $0 (${consecutiveZeros}/${ZERO_CONFIRM_COUNT}) — using cached $${lastKnownBalance.toFixed(2)}`);
      return lastKnownBalance;
    }

    consecutiveZeros = 0;
    lastKnownBalance = balance;
    return balance;
  } catch (err: unknown) {
    logger.error(`Failed to fetch USDC balance: ${errorMessage(err)}`);
    return lastKnownBalance > 0 ? lastKnownBalance : -1;
  }
}

File: poly_poly_bot/src/copy_trading/get_balance.py
"""USDC balance check with spurious-zero protection."""

from __future__ import annotations

from web3 import Web3

from src.config import CONFIG, get_private_key
from src.constants import ERC20_BALANCE_ABI, USDC_ADDRESS
from src.logger import logger
from src.utils import error_message

USDC_DECIMALS = 6

# Spurious-zero protection: cache last known balance and require
# multiple consecutive zero readings before reporting zero.
_ZERO_CONFIRM_THRESHOLD = 3
_last_known_balance: float | None = None
_consecutive_zeros: int = 0


def _get_wallet_address() -> str:
    """Derive wallet address from private key."""
    pk = get_private_key()
    w3 = Web3()
    account = w3.eth.account.from_key(f"0x{pk}")
    return account.address


def get_usdc_balance() -> float:
    """Read the USDC balance for the bot's proxy wallet (or main wallet).

    Returns:
        Balance in USDC (float). Returns cached value on RPC error.
        Returns -1.0 if no prior balance is available and RPC fails.

    Spurious-zero protection:
        If the RPC returns 0 but we previously had a non-zero balance,
        we require 3 consecutive zero readings before reporting zero.
    """
    global _last_known_balance, _consecutive_zeros

    wallet = CONFIG.proxy_wallet or _get_wallet_address()

    try:
        w3 = Web3(Web3.HTTPProvider(CONFIG.rpc_url))
        contract = w3.eth.contract(
            address=Web3.to_checksum_address(USDC_ADDRESS),
            abi=ERC20_BALANCE_ABI,
        )
        raw_balance: int = contract.functions.balanceOf(
            Web3.to_checksum_address(wallet)
        ).call()
        balance = raw_balance / (10 ** USDC_DECIMALS)

    except Exception as exc:
        logger.warn(f"RPC error fetching USDC balance: {error_message(exc)}")
        if _last_known_balance is not None:
            return _last_known_balance
        return -1.0

    # Spurious-zero protection
    if balance == 0.0 and _last_known_balance is not None and _last_known_balance > 0:
        _consecutive_zeros += 1
        if _consecutive_zeros < _ZERO_CONFIRM_THRESHOLD:
            logger.warn(
                f"Spurious zero balance? ({_consecutive_zeros}/{_ZERO_CONFIRM_THRESHOLD}), "
                f"returning cached {_last_known_balance:.2f}"
            )
            return _last_known_balance
        # Confirmed zero after threshold
        logger.warn(f"Balance confirmed zero after {_ZERO_CONFIRM_THRESHOLD} readings")

    if balance > 0:
        _consecutive_zeros = 0

    _last_known_balance = balance
    return balance
